### DISCUSSION TITLE
MaterialX libs in mayaHydra on OSX

### DISCUSSION THREAD

**Discussion by katpet** [Autodesk/maya-usd#3186] [Q&A]
Hi,
I have a hydra plug-in which links mayausd's MaterialX libraries. However, I'm running into trouble on OSX with the one called libMaterialXGenOslMayaUSD. The problem appears to be that the mayaHydra plug-in only links some of the MaterialX libs, so when my plug-in is loaded via the mayaHydra plug-in, the needed library isn't resolved. I can get it to work if I explicitly link my plug-in with an added rpath, like this:

` -rpath @executable_path/../../../../mayausd/maya2024/0.23.0_202303171117-8b326f9/mayausd/USD/lib/
`
But that's clearly awkward, since that may not be the correct location for users of the plug-in.

Could libMaterialXGenOslUSD.1.dylib be added to the mayaHydra link?

Here are the ones I see in there now:

```
otool -L /Applications/Autodesk/mayausd/maya2024/0.23.0_202303171117-8b326f9/mayausd/MayaUSD/lib/maya/mayaHydra.bundle|grep MaterialX
	@rpath/libMaterialXRenderMayaUSD.1.dylib (compatibility version 1.0.0, current version 1.38.5)
	@rpath/libMaterialXGenGlslMayaUSD.1.dylib (compatibility version 1.0.0, current version 1.38.5)
	@rpath/libMaterialXGenShaderMayaUSD.1.dylib (compatibility version 1.0.0, current version 1.38.5)
	@rpath/libMaterialXFormatMayaUSD.1.dylib (compatibility version 1.0.0, current version 1.38.5)
	@rpath/libMaterialXCoreMayaUSD.1.dylib (compatibility version 1.0.0, current version 1.38.5)
	@rpath/libMaterialXRenderMayaUSD.1.dylib (compatibility version 1.0.0, current version 1.38.5)
	@rpath/libMaterialXGenGlslMayaUSD.1.dylib (compatibility version 1.0.0, current version 1.38.5)
	@rpath/libMaterialXGenShaderMayaUSD.1.dylib (compatibility version 1.0.0, current version 1.38.5)
	@rpath/libMaterialXFormatMayaUSD.1.dylib (compatibility version 1.0.0, current version 1.38.5)
	@rpath/libMaterialXCoreMayaUSD.1.dylib (compatibility version 1.0.0, current version 1.38.5)
```

**Comment by lanierd-adsk** [comment]
Hi Katrin,
I am confused about your request. 
I understand you need to link with that lib but it doesn't make sense to me that we link with a library we don't need in maya-hydra, it will be ignored anyway if it is not used in by project.
About the context, is the plugin you are developing open source as well ?
Thank you.
Regards,
David

**Comment by lanierd-adsk** [comment]
More information about that rom Rakesh, one of our team member on maya-hydra at Autodesk :
"MayaHydra is not dealing with MaterialX files directly. 
The only practical use case at the moment is MaterialX through MayaUSD to process it for us. 
But MayaUSD is building against Mtlx already."
Regards,
David

**Reply by katpet** [reply]
My point was that since mayaHydra loads plug-ins that may need MaterialXGenOslUSD, it would be great if mayaHydra can link to that as well, so that when my plug-in is loaded by mayaHydra the lib will resolve. I realize mayaHydra itself doesn't use the lib, but the plug-ins it loads may. It doesn't appear that mayaHydra currently links with MaterialXGenOslUSD, given the otool output in my original post. This makes it my plug-in's problem to make sure that the lib can be found, which is difficult. I can make it work, but it involves adding an rpath guessing where to look for it.
